classlib/src/main/java/org/teavm/classlib/java/util/TRandom.java

```diff
@@ -75,15 +75,54 @@ public int nextInt(int n) {
     }
 
     public int nextInt(int origin, int bound) {
-        return origin + nextInt() * (bound - origin);
+        if (origin >= bound) {
+            throw new IllegalArgumentException();
+        }
+        int range = bound - origin;
+        if (range > 0) {
+            return nextInt(range) + origin;
+        } else {
+            while (true) {
+                int value = nextInt();
+                if (value >= origin && value < bound) {
+                    return value;
+                }
+            }
+        }
     }
 
     public long nextLong() {
         return ((long) nextInt() << 32) | nextInt();
     }
 
+    public long nextLong(long bound) {
```

Comment:
Would it make sense to use a similar implementation to `int nextInt(int n)` here instead of what was in the private method?

```
public long nextLong(long bound) {
        if (bound <= 0) {
            throw new IllegalArgumentException();
        }
        return (long) (nextDouble() * bound);
    }
```

Reply:
No, it's not enough of bytes (52 if I recall correctly) in mantissa of double to cover all possible longs. You're cutting most of possible results by this implementation.